File: mac-app/QuartzTeachers/Models/CourseActivity.swift
import Foundation

/// Which courses are doing something right now, across every window:
/// previewing (known via `PreviewLeases`) or publishing (recorded here).
///
/// Re-running the course setup rewrites a course's folders and files, so
/// actions like "Add Section…" must decline while one of the course's
/// sections is previewing or publishing — the setup run and the build
/// could stomp on each other's files.
@MainActor
enum CourseActivity {

    // MARK: - Types

    struct PublishRecord: Equatable {
        let folderPath: String
        let courseCode: String
        let sectionNumber: Int
    }

    // MARK: - Stored properties

    /// The publishes currently running, across all windows.
    private(set) static var activePublishes: [PublishRecord] = []

    // MARK: - Functions

    /// Records that a publish of one section has begun.
    static func beginPublish(folderPath: String, courseCode: String, sectionNumber: Int) {
        let record: PublishRecord = PublishRecord(
            folderPath: folderPath,
            courseCode: courseCode,
            sectionNumber: sectionNumber
        )
        activePublishes.append(record)
    }

    /// Records that a publish has finished, however it finished.
    static func endPublish(folderPath: String, courseCode: String, sectionNumber: Int) {
        let finished: PublishRecord = PublishRecord(
            folderPath: folderPath,
            courseCode: courseCode,
            sectionNumber: sectionNumber
        )
        var remaining: [PublishRecord] = []
        var didRemoveOne: Bool = false
        for existing in activePublishes {
            if existing == finished && !didRemoveOne {
                didRemoveOne = true
                continue
            }
            remaining.append(existing)
        }
        activePublishes = remaining
    }

    /// True while any section of the course is previewing or publishing.
    static func courseIsBusy(folderPath: String, courseCode: String) -> Bool {
        for lease in PreviewLeases.active {
            if lease.folderPath == folderPath && lease.courseCode == courseCode {
                return true
            }
        }
        for publish in activePublishes {
            if publish.folderPath == folderPath && publish.courseCode == courseCode {
                return true
            }
        }
        return false
    }

    /// Starts from nothing — for tests.
    static func reset() {
        activePublishes = []
    }
}

File: windows-app/Plantoir.Core/Models/CourseActivity.cs
namespace Plantoir.Core.Models;

/// <summary>
/// The cross-window answer to "is this course busy right now?" — previews
/// via PreviewLeases (held for the preview's whole life), publishes via the
/// begin/end records here. Adding a section re-runs the course setup, which
/// rewrites the course's folders and files while a preview or publish may be
/// mid-copy of those very files; the Add Section entry points decline while
/// their course is busy anywhere. Mirrors the mac app's CourseActivity
/// (row 104). The same code in a different working folder is a different
/// course, so every question carries the folder path.
/// </summary>
public static class CourseActivity
{
    private sealed record PublishRecord(string FolderPath, string CourseCode, int SectionNumber);

    private static readonly List<PublishRecord> _publishes = new();
    private static readonly object _gate = new();

    /// <summary>
    /// Record a publish for its whole life: dispose the token on EVERY exit
    /// path (success, failure, a build that never reached publishing).
    /// Disposing twice is harmless.
    /// </summary>
    public static IDisposable BeginPublish(string folderPath, string courseCode, int sectionNumber)
    {
        var record = new PublishRecord(folderPath, courseCode, sectionNumber);
        lock (_gate) _publishes.Add(record);
        return new PublishToken(record);
    }

    private sealed class PublishToken(PublishRecord record) : IDisposable
    {
        private PublishRecord? _record = record;
        public void Dispose()
        {
            if (_record is null) return;
            lock (_gate) _publishes.Remove(_record);
            _record = null;
        }
    }

    public static bool IsPreviewing(string folderPath, string courseCode) =>
        PreviewLeases.Active.Any(l => l.FolderPath == folderPath && l.CourseCode == courseCode);

    public static bool IsPublishing(string folderPath, string courseCode)
    {
        lock (_gate) return _publishes.Any(p => p.FolderPath == folderPath && p.CourseCode == courseCode);
    }

    /// <summary>
    /// True when an assistant is working on this course in another process.
    /// Unlike previews and publishes this is read from disk, because the MCP
    /// server has its own memory and neither side can see the other's.
    /// </summary>
    public static bool IsAssisting(string folderPath, string courseCode) =>
        Assist.WorkLease.IsHeld(folderPath, courseCode, Assist.WorkLease.Assisting);

    /// <summary>
    /// Another process is running a build of this course right now.
    ///
    /// The question Preview and Deploy ask before starting. Note what it does
    /// NOT ask: whether a conversation is open. A teacher previewing a section
    /// while asking the assistant to change it is the intended way to use both,
    /// and only the few seconds of an actual build are exclusive.
    /// </summary>
    public static bool IsBuildingElsewhere(string folderPath, string courseCode) =>
        Assist.WorkLease.IsHeld(folderPath, courseCode, Assist.WorkLease.Building);

    /// <summary>
    /// The short line naming what stands in the way of structural work on
    /// this course, or null when it is free.
    /// </summary>
    public static string? BusyReason(string folderPath, string courseCode)
    {
        bool previewing = IsPreviewing(folderPath, courseCode);
        bool publishing = IsPublishing(folderPath, courseCode);
        if (previewing && publishing) return "Available once preview and deploy complete";
        if (previewing) return "Available once preview completed";
        if (publishing) return "Available once deploy completed";
        // Said in the app's voice, naming the thing the teacher started rather
        // than the process that holds the lease.
        if (IsAssisting(folderPath, courseCode)) return "Available once you finish revising with Claude";
        return null;
    }

    public static void Reset()
    {
        lock (_gate) _publishes.Clear();
        PreviewLeases.Reset();
    }
}
